// gatsby-plugin-robots-txt/gatsby-node.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.onPostBuild = undefined;

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var onPostBuild = exports.onPostBuild = function () {
  var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(_ref, pluginOptions) {
    var graphql = _ref.graphql;

    var userOptions, defaultOptions, mergedOptions, _ref3, siteUrl, policy, sitemap, host, output, content, filename;

    return _regenerator2.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            userOptions = (0, _assign2.default)({}, pluginOptions);


            delete userOptions.plugins;

            defaultOptions = {
              output: '/robots.txt'
            };
            mergedOptions = (0, _assign2.default)({}, defaultOptions, userOptions);

            if (!(!mergedOptions.hasOwnProperty('host') || !mergedOptions.hasOwnProperty('sitemap'))) {
              _context.next = 11;
              break;
            }

            _context.next = 7;
            return runQuery(graphql, query);

          case 7:
            _ref3 = _context.sent;
            siteUrl = _ref3.site.siteMetadata.siteUrl;


            mergedOptions.host = siteUrl;
            mergedOptions.sitemap = _url2.default.resolve(siteUrl, 'sitemap.xml');

          case 11:
            policy = mergedOptions.policy, sitemap = mergedOptions.sitemap, host = mergedOptions.host, output = mergedOptions.output;
            _context.next = 14;
            return (0, _generateRobotstxt2.default)({
              policy: policy,
              sitemap: sitemap,
              host: host
            });

          case 14:
            content = _context.sent;
            filename = _path2.default.join(publicPath, output);
            _context.next = 18;
            return writeFile(filename, content);

          case 18:
            return _context.abrupt('return', _context.sent);

          case 19:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this);
  }));

  return function onPostBuild(_x, _x2) {
    return _ref2.apply(this, arguments);
  };
}();

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _url = require('url');

var _url2 = _interopRequireDefault(_url);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _generateRobotstxt = require('generate-robotstxt');

var _generateRobotstxt2 = _interopRequireDefault(_generateRobotstxt);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var publicPath = './public';
var query = '{\n  site {\n    siteMetadata {\n      siteUrl\n    }\n  }\n}\n';

function writeFile(file, data) {
  return new _promise2.default(function (resolve, reject) {
    _fs2.default.writeFile(file, data, function (err) {
      if (err) {
        reject(err);
      } else {
        resolve();
      }
    });
  });
}

function runQuery(handler, query) {
  return handler(query).then(function (res) {
    if (res.errors) {
      throw new Error(res.errors.join(', '));
    }

    return res.data;
  });
}
